### PR TITLE
Add debug card capabilities

### PR DESCRIPTION
**Pull request by yoavkatz**
Added debug_card utility to help debug cards
    
    At the core, added an ability to run a sequence operator
    only for a given number of steps.
    
    The debug_card utility iterates over all steps in a card and
    prints the for each step all the streams (splits) in the output
    and the first example in each stream.
    
    test_card(card,debug=True  )

### CODE REVIEW

File: prepare/cards/belebele.py [hunk 1]
```diff
@@ -182,5 +182,5 @@
         templates=CONTEXT_MMLU_TEMPLATES_NO_INTRO,
     )
     if lang == "acm_Arab":
-        test_card(card, tested_split="test")
+        test_card(card, demos_taken_from="test", debug=True)
```
Comment (matanor) [inline prepare/cards/belebele.py:185]: just checking, should commit with `debug=True`?

File: prepare/cards/sst2.py [hunk 2]
```diff
@@ -49,5 +49,5 @@
     templates="templates.one_sent_classification",
 )
 
-test_card(card)
+test_card(card, debug=False)
```
Comment (matanor) [inline prepare/cards/sst2.py:52]: again, just checking, if that's the default, should it be removed?

File: src/unitxt/operator.py [hunk 3]
```diff
@@ -176,16 +177,17 @@ def _process_single_stream(self, stream: Stream, stream_name: str = None) -> Str
...
             raise ValueError(
                 f"Stream '{stream_name}' can be in either apply_to_streams or dont_apply_to_streams not both."
             )
 
-        return (
-            self.apply_to_streams is None or stream_name in self.apply_to_streams
-        ) and stream_name not in self.dont_apply_to_streams
+        return (self.apply_to_streams is None or stream_name in self.apply_to_streams) and (
+            self.dont_apply_to_streams is None or stream_name not in self.dont_apply_to_streams
+        )
```
Comment (matanor) [inline src/unitxt/operator.py:190]: this part maybe deserves some u-tests (maybe open an issue for that?).

File: src/unitxt/operator.py [hunk 4]
```diff
@@ -351,22 +353,42 @@ def process(self, instance: dict, multi_stream: MultiStream) -> dict:
...
+
+    def set_max_steps(self, max_steps):
+        assert (
+            max_steps <= self.num_steps()
+        ), f"Max steps requested ({max_steps}) is larger than defined steps {self.num_steps()}"
+        assert max_steps >= 1, f"Max steps requested ({max_steps}) is less than 1"
+        self.max_steps = max_steps
+
+    def get_last_step_description(self):
+        last_step = self.max_steps - 1 if not self.max_steps is None else len(self.steps) - 1
```
Comment (matanor) [inline src/unitxt/operator.py:378]:
```suggestion
        last_step = self.max_steps - 1 if self.max_steps is not None else len(self.steps) - 1
```

Comment (matanor) [reply src/unitxt/operator.py:378]: (improves readability IMO)

File: src/unitxt/operator.py [hunk 5]
```diff
@@ -351,22 +353,42 @@ def process(self, instance: dict, multi_stream: MultiStream) -> dict:
...
+        description = str(self.steps[last_step])
+        return re.sub(r"\w+=None, ", "", description)
+
+    def _get_max_steps(self):
+        return self.max_steps if not self.max_steps is None else len(self.steps)
```
Comment (matanor) [inline src/unitxt/operator.py:383]:
```suggestion
        return self.max_steps if self.max_steps is not None else len(self.steps)
```

Comment (matanor) [reply src/unitxt/operator.py:383]: i think also just `self.max_steps if self.max_steps else len(self.steps)` should work

File: src/unitxt/splitters.py [hunk 6]
```diff
@@ -183,6 +183,10 @@ def process(self, instance: Dict[str, object], multi_stream: MultiStream) -> Dic
             raise Exception(f"Unable to fetch instances from '{self.source_stream}' to '{self.target_field}'")
 
 
+class CreateDemoSplit(SpreadSplit):
+    pass
```
Comment (matanor) [inline src/unitxt/splitters.py:187]: what is this part for?

File: src/unitxt/standard.py [hunk 7]
```diff
@@ -108,7 +108,7 @@ def prepare(self):
             sampler.set_size(self.num_demos)
 
             self.steps.append(
-                SpreadSplit(
+                CreateDemoSplit(
```
Comment (matanor) [inline src/unitxt/standard.py:111]: same as above, why the new class?

**Review by matanor — APPROVED** [review]
LGTM, left just a few minor comments

File: prepare/cards/sst2.py (hunk 2 above)
Comment (yoavkatz) [reply prepare/cards/sst2.py:52]: Left as an example.

File: src/unitxt/operator.py (hunk 3 above)
Comment (yoavkatz) [reply src/unitxt/operator.py:190]: There are tests for it.  See  test_stream_refiner.

File: src/unitxt/splitters.py (hunk 6 above)
Comment (yoavkatz) [reply src/unitxt/splitters.py:187]: To give a meaningful name to that part of the pipeline in debug mode.

**Pull request merged**